tools/checkescape/checkescape.go

```diff
@@ -0,0 +1,436 @@
+//
+// This is used to match against possible exemptions placed in the source.
+type LinePosition struct {
+	Filename string
+	Line     int
+}
+
+// String implements fmt.Stringer.String.
+func (e *Escape) String() string {
+	if e.Depth > 0 {
+		return fmt.Sprintf("%s (source position %s:%d)", e.Reason, e.Resolved.Filename, e.Resolved.Line)
+	}
+	return e.Reason
+}
+
+const (
+	interfaceInvoke = "interface: function invocation via interface"
+	unknownPackage  = "unknown: no package information available"
+	allocation      = "heap: call to runtime heap allocation"
+	builtin         = "builtin: call to runtime builtin"
+	closure         = "closure: closure allocation"
+	dynamicCall     = "dynamic: call via dynamic function"
+	internalError   = "error: internal error"
+)
+
+var hardReasons = []string{
+	allocation,
+	builtin,
+	closure,
+	internalError,
+}
+
+var softReasons = []string{
+	interfaceInvoke,
+	unknownPackage,
+	dynamicCall,
+}
+
+var allReasons = append(hardReasons, softReasons...)
+
+var escapeTypes = func() map[string]string {
+	result := make(map[string]string)
+	for _, r := range allReasons {
+		parts := strings.Split(r, ":")
+		result[parts[0]] = r // Key before ':'.
+	}
+	return result
+}()
+
+// run performs the analysis.
+func run(pass *analysis.Pass) (interface{}, error) {
+	pef := packageEscapeFacts{
+		Funcs: make(map[string][]Escape),
+	}
+	escapes := func(reason string, inst ssa.Instruction) []Escape {
+		p := pass.Fset.Position(inst.Pos())
+		return []Escape{
+			Escape{
+				Reason:   reason,
+				LocalPos: inst.Pos(),
+				Resolved: LinePosition{
+					Filename: p.Filename,
+					Line:     p.Line,
+				},
+			},
+		}
+	}
+	resolve := func(sub []Escape, pos token.Pos) (es []Escape) {
+		for _, e := range sub {
+			es = append(es, Escape{
+				Reason:   e.Reason,
+				LocalPos: pos,
+				Resolved: e.Resolved,
+				Depth:    e.Depth + 1,
+			})
+		}
+		return es
+	}
+	state := pass.ResultOf[buildssa.Analyzer].(*buildssa.SSA)
+
+	var loadFunc func(*ssa.Function) []Escape // Used below.
+
+	analyzeInstruction := func(inst ssa.Instruction) []Escape {
+		if call, ok := inst.(*ssa.Call); ok && call.Call.IsInvoke() {
+			// This is an interface dispatch.
+			return escapes(interfaceInvoke, inst)
+		} else if ok {
+			switch x := call.Call.Value.(type) {
+			case *ssa.Function:
+				if x.Pkg == nil {
+					// Can't resolve the package.
+					return escapes(unknownPackage, inst)
+				}
+
+				// Is this a local function? If yes, call the
+				// function to load the local function. The
+				// local escapes are the escapes found in the
+				// local function.
+				if x.Pkg.Pkg == pass.Pkg {
+					return resolve(loadFunc(x), inst.Pos())
+				}
+
+				// Recursively collect information from
+				// the other analyzers.
+				var imp packageEscapeFacts
+				if !pass.ImportPackageFact(x.Pkg.Pkg, &imp) {
+					// Is the package the sync/atomic package?
+					if x.Pkg.Pkg.Name() == "atomic" {
```

Comment:
Does the analyzer barf on the atomic package? (Because most of the calls are replaced with intrinsics?)

Reply:
No, we won't have any PackageFacts from the standard library. So we have to call those escapes as "unknown". I've effectively whitelisted the sync and sync/atomic packages here. I could also whitelist the syscall package, but it seems like that should be handled with an individual line tag.